### PR TITLE
Changing skip-service-worker flag to use-service-workers enum

### LINKED ISSUE #303: Only specific APIs should skip the fetch event when called within a service worker
Currently it is not clear whether resource fetches for notifications (e.g. icons) shown from a service worker should trigger the service worker onfetch handler. As a use case it makes sense to enable offlining these resources.

My understanding is that fetches initiated directly from a service worker are not passed through the service worker onfetch handler, which seems reasonable. However, what about the fetches triggered indirectly by actions initiated in the service worker, like showing a notification?

CC @annevk @jakearchibald @beverloo @slightlyoff

### PR DESCRIPTION
**Pull request by jakearchibald**
This fixes #303. I don't think there's anything else to do with this PR, but I need to make changes to SW to set this flag on particular APIs.

### CODE REVIEW

**Comment by jakearchibald** [comment]
Technically, this and skip-service-worker could become an enum of "none", "local" and "all", but it feels like that'd break specs already using it. Is that the right thing to do?

**Comment by jakearchibald** [comment]
Related SW PR https://github.com/w3c/ServiceWorker/pull/1025

**Comment by annevk** [comment]
I think I prefer the enum approach. There's not too many consumers as far as I know.

**Comment by jakearchibald** [comment]
Cool, I'll update the PR

**Comment by annevk** [comment]
Thanks. There's also #362 which relates to this flag somewhat.

File: fetch.bs [hunk 1]
```diff
@@ -728,11 +728,9 @@ explicitly set <a for=/>request</a>'s
 this flag set are subject to additional processing requirements.
 
 <p>A <a for=/>request</a> has an associated
-<dfn export>skip-service-worker flag</dfn>. Unless stated otherwise it is
-unset.
-
-<p>A <a for=/>request</a> has an associated <dfn export>skip-local-service-worker flag</dfn>. Unless
-stated otherwise it is unset.
+<dfn export>use-service-workers value</dfn>, that is "<code>all</code>",
```
Comment (annevk) [inline fetch.bs]: This needs `for=request` too. Seems we forgot that until now.

**Comment by annevk** [comment]
So where do we check for the `foreign` value? If we only ever set it, it seems like we could also reuse another value, no?

**Comment by annevk** [comment]
Do we have WPT tests around this?

**Comment by jakearchibald** [comment]
> So where do we check for the foreign value? If we only ever set it, it seems like we could also reuse another value, no?

Yeah, when it comes to performing foreign fetch we're only really concerned that the value *isn't* "none". Although I'm not sure what else could be reused.

> Do we have WPT tests around this?

No, I guess this is something I should learn and do 😄

**Comment by annevk** [comment]
> Yeah, when it comes to performing foreign fetch we're only really concerned that the value isn't "none".

And it cannot be "all" either I guess? I guess "foreign" is fine then. Can @mkruisselbrink have a look or is he still away?

**Comment by annevk** [comment]
As for the tests, the new rule is that they land together with the specification change. Let me know if you need help.

**Comment by annevk** [comment]
And only https://w3c.github.io/ServiceWorker/ seems to use the skip-service-worker flag currently. We should maybe prepare a PR there too so they can all land at the same time.

**Comment by jakearchibald** [comment]
https://github.com/w3c/ServiceWorker/pull/1025

File: fetch.bs [hunk 2]
```diff
@@ -728,8 +728,9 @@ explicitly set <a for=/>request</a>'s
 this flag set are subject to additional processing requirements.
 
 <p>A <a for=/>request</a> has an associated
-<dfn export>skip-service-worker flag</dfn>. Unless stated otherwise it is
-unset.
+<dfn export>use-service-workers value</dfn>, that is "<code>all</code>",
```
Comment (annevk) [inline fetch.bs]: What do you think about `<dfn export for=request>service-workers mode</dfn>`? I think that would be more consistent with what we have thus far.

Comment (jakearchibald) [inline fetch.bs]: SGTM, I wasn't happy with the naming

File: fetch.bs [hunk 3]
```diff
@@ -2793,16 +2794,14 @@ optional <i>CORS flag</i> and <i>CORS-preflight flag</i>, run these steps:
  <li><p>Let <var>actualResponse</var> be null.
 
  <li>
-  <p>If <var>request</var>'s <a>skip-service-worker flag</a> is unset, then run these
+  <p>If <var>request</var>'s <a>service-workers mode</a> is not "<code>none</code>", then run these
   substeps:
 
   <ol>
    <li>
     <p>If <var>request</var>'s <a for=request>client</a> is null or
```
Comment (mkruisselbrink) [inline fetch.bs]: Do we still need the client is null check, or is just relying on the service-workers mode enough now? And in particular it seems wrong that service-workers mode==foreign and client==null will now result in handle fetch being called.

File: fetch.bs [hunk 4]
```diff
@@ -728,8 +728,9 @@ explicitly set <a for=/>request</a>'s
 this flag set are subject to additional processing requirements.
 
 <p>A <a for=/>request</a> has an associated
-<dfn export>skip-service-worker flag</dfn>. Unless stated otherwise it is
-unset.
+<dfn for=request export>service-workers mode</dfn>, that is "<code>all</code>",
```
Comment (mkruisselbrink) [inline fetch.bs]: Would it make sense to add some non-normative text somewhere to explain what this mode actually means? Of course you can figure that out by reading the algorithms, but I find the non-normative notes that many of the requests attributes have to be quite helpful in quickly figuring out what a particular thing is for.

**Comment by mkruisselbrink** [comment]
> > Do we have WPT tests around this?
>
> No, I guess this is something I should learn and do

Hmm, I should probably also work on actually upstreaming the existing foreign fetch tests we have in blink.  Although not sure what WPT tests would be desired for this specific change, as this change by itself doesn't seem to have any behavioral changes?

File: fetch.bs (hunk 4 above)
Comment (annevk) [inline fetch.bs]: Yeah, sounds good.

File: fetch.bs (hunk 3 above)
Comment (annevk) [inline fetch.bs]: I think that relying on service-workers mode is enough. At least, I thought the idea was that having a client no longer mattered.

**Comment by annevk** [comment]
>  Although not sure what WPT tests would be desired for this specific change, as this change by itself doesn't seem to have any behavioral changes?

The fact that certain fetches from the service worker can now invoke the service worker is a normative change that will result in behavioral changes.

**Comment by jakearchibald** [comment]
I agree this is a change that needs tests. Will pick this up after the Christmas break.

**Comment by jakearchibald** [comment]
Tests: https://github.com/jakearchibald/web-platform-tests/commit/0fa7353b8412182feac56c350542b3318ebd53ab

@annevk would be grateful for a review. Be kind!

**Comment by annevk** [comment]
@jakearchibald probably best to create a PR on web-platform-tests for review. Overall I think it's fine, but since you need permissions you'll have to use the -manual suffix I think.

**Comment by jakearchibald** [comment]
https://github.com/w3c/web-platform-tests/pull/4518

File: fetch.bs (hunk 4 above)
Comment (jakearchibald) [inline fetch.bs]: Done

**Comment by jakearchibald** [comment]
@annevk tests merged. Happy to merge this?

File: fetch.bs [hunk 5]
```diff
@@ -705,8 +705,27 @@ explicitly set <a for=/>request</a>'s
...
+"<code>foreign</code>", or "<code>none</code>". Unless stated otherwise it is
+"<code>all</code>".
+
+<div class=note>
+ This determines which service workers will receive a {{fetch!!event}} event for this fetch.
```
Comment (annevk) [inline fetch.bs]: This should start with a `<p>`.

**Comment by annevk** [comment]
One more question, in a commit you stated #362 is fixed by this, but that seems unlikely as service workers are skipped for redirects with this change. So are we leaving that for that issue?

Could you write a commit message for this?

**Comment by jakearchibald** [comment]
@annevk uhhhh good catch. Sorry about that. As far as I can tell it just needed that "none" changed to "foreign".

**Comment by jakearchibald** [comment]
Commit message:

```
Add "service-workers mode" flag to allow particular fetches within a service worker to trigger "fetch" events.

Fixes: #303, #362.
Related service worker PR: https://github.com/w3c/ServiceWorker/pull/1025.
Tests: https://github.com/w3c/web-platform-tests/pull/4518.
```

**Comment by annevk** [comment]
Thanks! (Adjusted the message a bit to fit style guidelines.)

**Pull request merged**